src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -274,40 +293,100 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
             }
             constantSourcesBuilder.Free();
 
-            // Run a syntax walk if any of the generators requested it
-            if (receiverCount > 0)
+
+            // PROTOTYPE(source-generators): we don't need to run at all if none of the inputs have changes.
+            var driverStateBuilder = new DriverStateTable.Builder(_state.StateTable, cancellationToken);
+            driverStateBuilder.SetInputState(SharedInputNodes.Compilation, NodeStateTable<Compilation>.WithSingleItem(compilation, EntryState.Added));
+            driverStateBuilder.SetInputState(SharedInputNodes.AnalzerConfigOptions, NodeStateTable<AnalyzerConfigOptionsProvider>.WithSingleItem(_state.OptionsProvider, EntryState.Added));
+            driverStateBuilder.SetInputState(SharedInputNodes.ParseOptions, NodeStateTable<ParseOptions>.WithSingleItem(_state.ParseOptions, EntryState.Added));
+
+            // build the updated node for the syntax trees
+            PooledHashSet<SyntaxTree> syntaxTrees = PooledHashSet<SyntaxTree>.GetInstance();
+            foreach (var tree in compilation.SyntaxTrees)
             {
-                foreach (var tree in compilation.SyntaxTrees)
+                syntaxTrees.Add(tree);
+            }
+
+            var previousTrees = state.StateTable.GetStateTable<SyntaxTree>(SharedInputNodes.SyntaxTrees);
+            var newTrees = new NodeStateTable<SyntaxTree>.Builder();
+            foreach ((var tree, var treeState) in previousTrees)
+            {
+                newTrees.AddEntriesFromPreviousTable(previousTrees, syntaxTrees.Contains(tree) ? EntryState.Cached : EntryState.Removed);
+                syntaxTrees.Remove(tree);
+            }
+            foreach (var tree in syntaxTrees)
+            {
+                newTrees.AddEntries(ImmutableArray.Create(tree), EntryState.Added);
+            }
+            var syntaxTreeTable = newTrees.ToImmutableAndFree();
+            driverStateBuilder.SetInputState(SharedInputNodes.SyntaxTrees, syntaxTreeTable);
+            syntaxTrees.Free();
+
+            // PROTOTYPE(source-generators): Can we do this inline as we update the trees above?
+            if (syntaxInputNodes.Count > 0 || receiverCount > 0)
+            {
+                var builders = syntaxInputNodes.SelectAsArray(n => n.GetBuilder(_state.StateTable));
```

Comment:
>var builders [](http://example.com/codeflow?start=16&length=12)

Could `builders` be a pooled `ArrayBuilder<ISyntaxTransformBuilder>` rather than `ImmutableArray<ISyntaxTransformBuilder>`?